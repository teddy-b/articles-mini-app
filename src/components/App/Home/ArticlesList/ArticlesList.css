.articles-list {
  width: 40%;
}

.articles-list h2 {
  margin-left: 1em
}

.articles-list .pagination {
  margin-top: 1em;
  margin-bottom: 2em;
  text-align: center;
}
