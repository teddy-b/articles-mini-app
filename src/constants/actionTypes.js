export const LOAD_ARTICLES_START = 'LOAD_ARTICLES_START';
export const LOAD_ARTICLES_SUCCESS = 'LOAD_ARTICLES_SUCCESS';
export const LOAD_ARTICLES_FAIL = 'LOAD_ARTICLES_FAIL';

export const SELECT_ARTICLE = 'SELECT_ARTICLE';

export const LOAD_COMMENTS_START = 'LOAD_COMMENTS_START';
export const LOAD_COMMENTS_SUCCESS = 'LOAD_COMMENTS_SUCCESS';
export const LOAD_COMMENTS_FAIL = 'LOAD_COMMENTS_FAIL';

export const LOAD_REPLIES_START = 'LOAD_REPLIES_START';
export const LOAD_REPLIES_SUCCESS = 'LOAD_REPLIES_SUCCESS';
export const LOAD_REPLIES_FAIL = 'LOAD_REPLIES_FAIL';

export const ADD_COMMENT_START = 'ADD_COMMENT_START';
export const ADD_COMMENT_SUCCESS = 'ADD_COMMENT_SUCCESS';
export const ADD_COMMENT_FAIL = 'ADD_COMMENT_FAIL';

export const ADD_REPLY_START = 'ADD_REPLY_START';
export const ADD_REPLY_SUCCESS = 'ADD_REPLY_SUCCESS';
export const ADD_REPLY_FAIL = 'ADD_REPLY_FAIL';

export const NEXT_PAGE = 'NEXT_PAGE';
export const PREV_PAGE = 'ADD_PREV_PAGEREPLY_FAIL';
