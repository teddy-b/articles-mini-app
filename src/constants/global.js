export const ARTICLES_PER_PAGE = 4;

export default ARTICLES_PER_PAGE;
