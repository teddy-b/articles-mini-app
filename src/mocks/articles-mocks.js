export const articlesMock = [
  {
    title: 'CERN experiment discovers five new particles',
    text: 'A new group of particles which have long been hiding in plain sight have finally been discovered thanks to the incredibly sensitive LHCb experiment at CERN. LHCb, otherwise known as the Large Hadron Collider beauty experiment, is one of seven experiments collecting data at the Large Hadron Collider at CERN. UK participation in the experiment is funded by STFC.',
    imageUrl: 'http://www.stfc.ac.uk/stfc/cache/file/4821074D-0D31-4964-9318CA0AE81E0639.jpg',
    id: 0,
    createdAt: 1530698004815,
    commentsCount: 8
  },
  {
    title: 'Show HN: Marta – A fast and minimalistic file manager for macOS',
    text: 'File Manager for macOS. Native. Minimalistic. Fast. ',
    imageUrl: 'https://marta.yanex.org/data/images/dark.png',
    id: 1,
    createdAt: 1530698004815,
    commentsCount: 7
  },
  {
    title: 'Tears',
    text: 'So humans are a peculiar species. You probably kne…specific traits and behaviors are uniquely human.',
    imageUrl: 'http://www.meltingasphalt.com/wp-content/uploads/2014/12/weeping_equation.png',
    id: 2,
    createdAt: 1530698004815,
    commentsCount: 0
  }
];

export default articlesMock;
