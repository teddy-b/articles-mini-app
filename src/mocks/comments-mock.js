export const commentsMock = [
  {
    articleId: 0,
    author: 'grabcocque',
    createdAt: 1530698004815,
    id: 0,
    parentCommentId: null,
    repliesCount: 0,
    text: 'I have to admit, I rather admire their putting a brave face on things, with their mildly-too-insistent claim that the LHC is vital for finding new physics despite the increasing likelihood it\'s not going to find any. In that sense the LHC is a political failure, because it has failed at its primary political job, which is to make the argument for an even bigger collider.'
  },
  {
    articleId: 0,
    author: 'teddythetwig',
    createdAt: 1530698004815,
    id: 1,
    parentCommentId: null,
    repliesCount: 2,
    text: 'I assume these particles were already predicted as part of the Standard Model?'
  },
  {
    articleId: 0,
    author: 'guitarbill',
    createdAt: 1530698004815,
    id: 6,
    parentCommentId: null,
    repliesCount: 0,
    text: 'AFAIK, the experimentally determined rest masses of these excited states/particles (same thing, different was of looking at it) agree with the calculated ones well. So yes, they were predicted. No surprises sadly; of course it\'s still a huge achievement!'
  }
];

export default commentsMock;
