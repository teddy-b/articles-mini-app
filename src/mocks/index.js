export * from './articles-mocks';
export * from './comments-mock';
export * from './replies-mock';
export * from './replies-replies-mock';
