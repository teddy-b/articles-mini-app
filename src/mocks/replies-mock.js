export const repliesMock = [
  {
    articleId: 0,
    author: 'brutuscat',
    createdAt: 1530698004815,
    id: 2,
    parentCommentId: 1,
    repliesCount: 0,
    text: 'Professor Tim Gershon, Professor of Physics at University of Warwick and UK spokesperson for the LHCb experiment: '
  },
  {
    articleId: 0,
    author: 'gpderetta',
    createdAt: 1530698004815,
    id: 3,
    parentCommentId: 1,
    repliesCount: 2,
    text: 'yes, and note that these are not fundamental particles (like the Higgs was for example), but composite particles of yet another combination of the fundamental quarks. The SM predicts the existence of hundreds (thousands?) of these.'
  }
];

export default repliesMock;
