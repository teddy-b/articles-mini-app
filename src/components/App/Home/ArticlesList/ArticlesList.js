import React from 'react';

import ArticleItem from './ArticleItem';
import { ARTICLES_PER_PAGE } from '../../../../constants/global';

import './ArticlesList.css';

const ArticlesList = ({
  articles,
  articlesCount,
  onLoadComments,
  onNextPage,
  onPrevPage,
  onSelectArticle,
  page,
  selectedArticleId
}) => {
  const start = page * ARTICLES_PER_PAGE;
  const end = start + ARTICLES_PER_PAGE;
  const articlesToShow = articles.slice(start, end);
  return (
    <div className="articles-list">
      <h2>Articles list</h2>
      {articlesToShow.map(article => (
        <ArticleItem
          key={article.id}
          {...article}
          onLoadComments={onLoadComments}
          onSelectArticle={onSelectArticle}
          selected={article.id === selectedArticleId}
        />
      ))}
      <div className="pagination">
        {page !== 0 && (
          <button type="button" onClick={onPrevPage}>Load Prev</button>
        )}
        {end < articlesCount && (
          <button type="button" onClick={onNextPage}>Load Next</button>
        )}
      </div>
    </div>
  );
};

export default ArticlesList;
