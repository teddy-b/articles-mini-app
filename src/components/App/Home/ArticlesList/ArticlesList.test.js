import React from 'react';
import renderer from 'react-test-renderer';

import ArticlesList from './ArticlesList';
import { ARTICLES_PER_PAGE } from '../../../../constants/global';
import { articlesMock } from '../../../../mocks';

jest.mock('./ArticleItem', () => () => 'ArticleItem');

describe('ArticlesList', () => {
  const props = {
    articles: [],
    articlesCount: 0,
    onLoadComments: jest.fn(),
    onNextPage: jest.fn(),
    onPrevPage: jest.fn(),
    onSelectArticle: jest.fn(),
    page: 0,
    selectedArticleId: null
  };

  it('renders correctly', () => {
    const component = renderer.create(
      <ArticlesList
        {...props}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with articles', () => {
    const component = renderer.create(
      <ArticlesList
        {...props}
        articles={articlesMock}
        articlesCount={articlesMock.length}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly load next button', () => {
    const component = renderer.create(
      <ArticlesList
        {...props}
        articles={articlesMock}
        articlesCount={ARTICLES_PER_PAGE + 1}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly load next and load prev button', () => {
    const component = renderer.create(
      <ArticlesList
        {...props}
        articles={articlesMock}
        articlesCount={2 * ARTICLES_PER_PAGE + 1}
        page={1}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly last page', () => {
    const component = renderer.create(
      <ArticlesList
        {...props}
        articles={articlesMock}
        articlesCount={ARTICLES_PER_PAGE + 1}
        page={1}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
