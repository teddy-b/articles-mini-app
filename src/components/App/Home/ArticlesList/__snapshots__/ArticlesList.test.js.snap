// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticlesList renders correctly 1`] = `
<div
  className="articles-list"
>
  <h2>
    Articles list
  </h2>
  <div
    className="pagination"
  />
</div>
`;

exports[`ArticlesList renders correctly last page 1`] = `
<div
  className="articles-list"
>
  <h2>
    Articles list
  </h2>
  <div
    className="pagination"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Load Prev
    </button>
  </div>
</div>
`;

exports[`ArticlesList renders correctly load next and load prev button 1`] = `
<div
  className="articles-list"
>
  <h2>
    Articles list
  </h2>
  <div
    className="pagination"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Load Prev
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      Load Next
    </button>
  </div>
</div>
`;

exports[`ArticlesList renders correctly load next button 1`] = `
<div
  className="articles-list"
>
  <h2>
    Articles list
  </h2>
  ArticleItem
  ArticleItem
  ArticleItem
  <div
    className="pagination"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Load Next
    </button>
  </div>
</div>
`;

exports[`ArticlesList renders correctly with articles 1`] = `
<div
  className="articles-list"
>
  <h2>
    Articles list
  </h2>
  ArticleItem
  ArticleItem
  ArticleItem
  <div
    className="pagination"
  />
</div>
`;
